sui_programmability/examples/nft_standard/sources/collection.move

```diff
@@ -0,0 +1,87 @@
+module nft_standard::collection {
+    struct Collection<phantom T, M: store> has key, store {
+        id: UID,
+        // Standard fields that every collection needs go here.
+        // Pulling a few obvious ones from https://github.com/Origin-Byte/nft-protocol/blob/main/sources/collection/collection.move and
+        // https://github.com/suiet/standard/blob/main/docs/nft/collections.md
+        // but more will be needed
+        /// Address that created this collection
+        creator: address,
+        /// Name of the collection. TODO: should this just be T.name?
+        name: String,
+        /// Description of the collection
+        description: String,
+        /// The maximum number of instantiated NFT objects. Use U64_MAX if there is no max
+        total_supply: u64,
+        // ... more standard fields
+        /// Custom metadata outside of the standard fields goes here
+        custom_metadata: M,
+    }
+
+    /// Proof that the given NFT is one of the limited `total_supply` NFT's in `Collection`
+    struct CollectionProof has store {
+        collection_id: ID
+    }
+
+    /// Grants the permission to mint `num_remaining` NFT's of type `T`
+    /// The sum of `num_remaining` for all mint caps + the number of instantiated
+    /// NFT's should be equal to `total_supply`.
+    /// This is a fungible type to support parallel minting, giving a buyer the permission
+    /// mint themselves, allowing multiple parties to mint, and so on.
+    struct MintCap<phantom T> has key, store {
+        id: UID,
+        /// ID of the collection that this MintCap corresponds to
+        collection: ID,
+        /// Number of NFT's this cap can mint
+        num_remaining: u64,
+    }
+
+    /// Grants the permission to mint `RoyaltyReceipt`'s for `T`.
+    /// Receipts are required when paying for NFT's
+    struct RoyaltyCap<phantom T> has key, store {
```

Comment:
Looking at this with a fresh perspective: perhaps what we could do is to have a method in `T` which can create a new `RoyaltyCap` at will for the admin. Then, the admin - after they verified that the given trading facilitator contract adheres to royalties - can give this object _to_ that contract. 

This means `RoyaltyCap` effectively acts as a whitelist.

Reply:
Therefore one could either go via the `T::buy_nft` royalty collection fn or be trusted by the collection admin by getting `RoyaltyCap` from them.